File: 784. Letter Case Permutation.cpp
class Solution {
public:    
    vector<string> letterCasePermutation(string S) {
        vector<string> res{S};
        for(int i=0; i<S.size(); i++) {
            vector<string> new_res;
            vector<string>::iterator it;
            for(it = res.begin(); it != res.end(); it++) {
                if(isalpha((*it)[i])){
                    new_res.push_back((*it).substr(0, i)+char(tolower((*it)[i]))+(*it).substr(i+1, S.size()));   
                    new_res.push_back((*it).substr(0, i)+char(toupper((*it)[i]))+(*it).substr(i+1, S.size())); 
                }
                else
                    new_res.push_back(*it);
            }
            res = new_res;
        }
        return res;
    }
};

File: string/784. Letter Case Permutation.py
"""
iteration, pretty fast
"""
class Solution(object):
    def letterCasePermutation(self, S):
        """
        :type S: str
        :rtype: List[str]
        """
        res = [""]
        for s in S:
            if s.isalpha():
                res = [i+j for i in res for j in (s.upper(), s.lower())]
            else:
                res = [i+s for i in res]
        return res

"""
recursion: super slow
"""
class Solution:
    def letterCasePermutation(self, S: str) -> List[str]:
        res = []
        self.helper(S, 0, "", res)
        return res
        
    def helper(self, S, index, path, res):
        if len(path) == len(S):
            res.append(path)
            return
        for i in range(index, len(S)):
            if S[i].isalpha():
                self.helper(S, i+1, path + S[i].upper(), res)
                self.helper(S, i+1, path + S[i].lower(), res)
            else:
                self.helper(S, i+1, path + S[i], res)
